### ISSUE TITLE
Support for Option in Multimapping

### ISSUE THREAD

**Issue by tronta** [nolanderc/rust-postgres-query#8]
Your crate is really helpful.
Do you plan support for Option in Multi Mapping as well? That would be great.
eg make the author optional:
```
#[derive(FromSqlRow)]
#[row(split)]
struct Book {
    #[row(split = "id")]
    id: i32,
    title: String,
    genre: String,

    #[row(flatten, split = "id")]
    author: Option<Author>,
}

#[derive(FromSqlRow)]
struct Author {
    id: i32,
    name: String,
    birthyear: i32,
}
```

**Comment by nolanderc** [comment]
Absolutely! I had totally forgotten about adding implementations for types in the standard library, thanks for reminding me 😄

**Comment by nolanderc** [comment]
Hmm, turns out it will be a bit more troublesome than I initially thought. The problem is with splits. Currently, if a row is partitioned based on splits and a specific split couldn't be made, we return an error. In the example you gave, and the `id` row was missing, that split could no longer be made, so we would throw an error. 

To fix this we would need some concept of an optional or fallible split. I will have to look into it, but expect that it will take a while.

**Comment by tronta** [comment]
I'm not an expert at all, but if I do the join and the FK is null, then the columns still exist, but all their values are `null`. 
As I said, I'm not an expert. Thanks a lot for looking into this.

**Comment by nolanderc** [comment]
Oh, I'm sorry, I was the one doing the misunderstanding. I interpreted `Option<T>` to mean: "there might be some columns here describing an author" and not "the columns here might not contain valid values". The latter case I have already implemented locally, I can push an update as soon as it is ready!

**Comment by tronta** [comment]
Oh, great. Just to double check (sorry for my poor problem description). Currently I get an error 
`the trait postgres_query::FromSqlRow is not implemented for std::option::Option<Author>`
So in my thinking, there might be data for an author or not. If no data/all values are `null` author would be `None`
Would this be solved with that?

**Comment by nolanderc** [comment]
Yeah, kind of. We'll have to figure out the specifics of what it means to return a `None`: do we want it to mean "failed to extract the value because some columns were null" or "an error occurred when extracting the value". 

The former is probably closer to what you are looking for, but I'm trying to think of a case when you'd want some other behavior. The latter is what is currently implemented in the pull request, mostly because it was easier and didn't require any breaking changes. As a proof of concept it works fine, but the problem is that all errors get silenced, which isn't the best debugging experience.

~~Either way, for this feature to work we would need some way to attach more information to an error to determine if it was a 'soft' error that could be converted into `None` or a 'hard' error that needs to be propagated upwards. Edit: changing the error type would be a breaking change and requires bumping the SemVer to 0.4.0~~

Edit: managed to figure it out without changing the error type

**Comment by nolanderc** [comment]
This feature is now available as part of postgres-query v0.3.3

**Comment by tronta** [comment]
Great. 👍  works now.
